llama_index/llms/bedrock_utils.py

```diff
@@ -61,38 +61,98 @@
+    def max_tokens_key(self) -> str:
+        ...
+
+    @abstractmethod
+    def get_text_from_response(self, response: dict) -> str:
+        ...
+
+    def get_text_from_stream_response(self, response: dict) -> str:
+        return self.get_text_from_response(response)
+
+    def get_request_body(self, prompt: str, inference_parameters: dict) -> dict:
+        return {"prompt": prompt, **inference_parameters}
+
+    messages_to_prompt: Optional[Callable[[Sequence[ChatMessage]], str]] = None
+    completion_to_prompt: Optional[Callable[[str], str]] = None
+
+
+class AmazonProvider(Provider):
+    max_tokens_key = "maxTokenCount"
+
+    def get_text_from_response(self, response: dict) -> str:
+        return response["results"][0]["outputText"]
+
+    def get_text_from_stream_response(self, response: dict) -> str:
+        return response["outputText"]
+
+    def get_request_body(self, prompt: str, inference_parameters: dict) -> dict:
+        return {
+            "inputText": prompt,
+            "textGenerationConfig": {**inference_parameters},
+        }
+
+
+class Ai21Provider(Provider):
+    max_tokens_key = "maxTokens"
+
+    def get_text_from_response(self, response: dict) -> str:
+        return response["completions"][0]["data"]["text"]
+
+
+def completion_to_anthopic_prompt(completion: str) -> str:
+    return messages_to_anthropic_prompt(prompt_to_messages(completion))
+
+
+class AnthropicProvider(Provider):
+    max_tokens_key = "max_tokens_to_sample"
+
+    def __init__(self) -> None:
+        self.messages_to_prompt = messages_to_anthropic_prompt
+        self.completion_to_prompt = completion_to_anthopic_prompt
+
+    def get_text_from_response(self, response: dict) -> str:
+        return response["completion"]
+
+
+class CohereProvider(Provider):
+    max_tokens_key = "max_tokens"
+
+    def get_text_from_response(self, response: dict) -> str:
+        return response["generations"][0]["text"]
+
+
+class MetaProvider(Provider):
+    max_tokens_key = "max_gen_len"
 
-PROVIDER_STREAM_RESPONSE_LOADER = {
-    "amazon": lambda x: x["outputText"],
-    "anthropic": lambda x: x["completion"],
-    "meta": lambda x: x["generation"],
+    def __init__(self) -> None:
+        self.messages_to_prompt = messages_to_llama_prompt
+        self.completion_to_prompt = completion_to_llama_prompt
+
+    def get_text_from_response(self, response: dict) -> str:
+        return response["generation"]
+
+
+PROVIDERS = {
+    "amazon": AmazonProvider(),
+    "ai21": Ai21Provider(),
+    "anthropic": AnthropicProvider(),
+    "cohere": CohereProvider(),
+    "meta": MetaProvider(),
 }
-logger = logging.getLogger(__name__)
 
 
-def bedrock_model_to_param_name(model: str, param_name: str) -> str:
-    provider = model.split(".")[0]
-    model_params = PROVIDER_SPECIFIC_PARAM_NAME[provider]
-    return model_params[param_name]
+def get_provider(model: str) -> Provider:
+    provider_name = model.split(".")[0]
+    return PROVIDERS[provider_name]
```

Comment:
should we raise an error in case the provider is not defined ? (and so provide a more meaningful error message to the user ?)

Reply:
probably a good idea

Reply:
done

Reply:
done